Rollbar/RollbarPayloadTruncator.m

```diff
@@ -0,0 +1,271 @@
+        }];
+    } else if ([obj isKindOfClass:[NSSet class]]) {
+        //recurse the collection obj's items:
+        [obj enumerateObjectsUsingBlock: ^(id item, BOOL *stop) {
+            [RollbarPayloadTruncator itereateObjectStructure:item
+                                       whileTuncatingStrings:stringBytesLimit
+             ];
+        }];
+    } else {
+        //nothing really...
+    }
+}
+
++(BOOL)isMutable:(id)str {
+    NSString *copy = [str copy];
+    return copy != str;
+}
+
++(BOOL)truncatePayload:(NSMutableDictionary*)payload
+          toTotalBytes:(unsigned long) payloadLimit
+       byReducingItems:(NSString*)pathToItems
+     keepingHeadsCount:(unsigned long)headsCount
+     keepingTailsCount:(unsigned long)tailsCount {
+    
+    if (![RollbarPayloadTruncator isTruncationNeeded:payload forLimit:payloadLimit]) {
+        return FALSE;  //payload is small enough, no need to truncate further...
+    }
+    
+    NSMutableArray *items = [payload mutableArrayValueForKeyPath:pathToItems];
+    if (items.count <= (headsCount + tailsCount)) {
+        return TRUE;
+    }
+    
+    unsigned long totalItemsToRemove = items.count - headsCount - tailsCount;
+    [items removeObjectsInRange:NSMakeRange(headsCount, totalItemsToRemove)];
+    
+    return TRUE;
+}
+
++(BOOL)isTruncationNeeded:(NSMutableDictionary*)payload forLimit:(unsigned long)payloadBytesLimit {
+    
+    NSData *jsonPayload = [NSJSONSerialization dataWithJSONObject:payload
+                                                          options:0
+                                                            error:nil
+                                                             safe:true
+                           ];
+    return (payloadBytesLimit < jsonPayload.length);
+}
+
++(unsigned long)measureTotalEncodingBytes:(NSString*)string {
+    
+    return [RollbarPayloadTruncator measureTotalEncodingBytes:string
+                                                usingEncoding:(NSUTF8StringEncoding)
+            ];
+}
+
+
+
+
++(unsigned long)measureTotalEncodingBytes:(NSString*)string
+                            usingEncoding:(NSStringEncoding)encoding {
+
+    NSUInteger totalBytes = [string lengthOfBytesUsingEncoding:encoding];
+    
+    return totalBytes;
+}
+
++(NSString*)truncateString:(NSString*)inputString
+              toTotalBytes:(unsigned long)totalBytesLimit {
+    
+    unsigned long currentStringEncoodingBytes =
+        [RollbarPayloadTruncator measureTotalEncodingBytes:inputString];
+    if (currentStringEncoodingBytes <= totalBytesLimit)
+    {
+        // no need to truncate:
+        return inputString;
+    }
+    
+    NSString *ellipsis = @"...";
+    unsigned long totalEllipsisEncodingBytes =
+        [RollbarPayloadTruncator measureTotalEncodingBytes:ellipsis];
+    unsigned long bytesToRemove =
+        currentStringEncoodingBytes - totalBytesLimit + totalEllipsisEncodingBytes;
+
+    NSMutableString *result =
```

Comment:
In my testing of this it seems to truncate too much.

Reply:
did you test by running the testPayloadTruncation unit test? if yes, it intentionally forces an aggressive truncation since it sets payload limit to 20 bytes.

Reply:
I ran this code in isolation with some specific examples, I think there is something off by 1. If you set `totalBytesLimit` to 5 and pass in the string `ABCDEF` the result here is `A...`. It should probably be `AB...` to match the byte limit. Dropping the `F` at end results in `ABCDE` which is what I expect. I think `substringToIndex:inputString.length - bytesToRemove - 1` should be `substringToIndex:inputString.length - bytesToRemove` When I initially left this comment I thought it was truncating even more than off by 1, so not that big of a deal.

Reply:
oh, yes:
"substringToIndex:
Returns a new string containing the characters of the receiver up to, but not including, the one at a given index."

removing "-1"...
thanks!